sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -539,3 +539,79 @@ case class StringToMap(text: Expression, pairDelim: Expression, keyValueDelim: E
 
   override def prettyName: String = "str_to_map"
 }
+
+/**
+ * Adds/replaces field in struct by name.
+ */
+case class WithFields(
+    structExpr: Expression,
+    names: Seq[String],
+    valExprs: Seq[Expression]) extends Expression {
+
+  assert(names.length == valExprs.length)
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (!structExpr.dataType.isInstanceOf[StructType]) {
+      TypeCheckResult.TypeCheckFailure(
+        "struct argument should be struct type, got: " + structExpr.dataType.catalogString)
+    } else {
+      TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  override def children: Seq[Expression] = structExpr +: valExprs
+
+  private lazy val addOrReplaceExprs = names.zip(valExprs)
+
+  override def dataType: StructType = {
```

Comment:
Can this simply be `evalExpr.dataType`? then we can avoid calling `addOrReplace` and doing the name lookup twice.

Reply:
We can't do this. `evalExpr.dataType` will **always** return a nullable datatype with my most recent changes.

Reply:
Just to be clear, even after reverting the codegen changes, this is still not possible.  
The main issue is with nullable structs. 
When we call `WithFields` on a nullable struct, we want the **existing** fields inside the struct to maintain the same nullability as before. Because we use `GetStructField` to extract the existing fields we won't get this behaviour as `GetStructField` returns a nullable dataType if the struct that it's being called upon is nullable. 
The `withField should add field to null struct` test is a good example of a test that will fail with your suggested change: 
```
StructType(StructField(a,StructType(StructField(a,IntegerType,true), StructField(b,IntegerType,true), StructField(c,IntegerType,true), StructField(d,IntegerType,false)),true)) 
did not equal 
StructType(StructField(a,StructType(StructField(a,IntegerType,false), StructField(b,IntegerType,true), StructField(c,IntegerType,false), StructField(d,IntegerType,false)),true))
```

Reply:
resolved due to comment [here](https://github.com/apache/spark/pull/27066#discussion_r448719460).